src/mmds/src/token.rs

```diff
@@ -0,0 +1,583 @@
+    fn decrypt_expiry(&self, ciphertext: &[u8], iv: &[u8]) -> Result<u64, Error> {
+        // Create Nonce object from initialization vector.
+        let nonce = Nonce::from_slice(iv);
+        // Decrypt expiry as vector of bytes from ciphertext.
+        let expiry_as_bytes = self
+            .cipher
+            .decrypt(nonce, ciphertext)
+            .map_err(|_| Error::ExpiryExtraction)?;
+        // Get expiry value in seconds from bytes.
+        let expiry: u64 =
+            bincode::deserialize(&expiry_as_bytes).map_err(|_| Error::ExpiryExtraction)?;
+
+        Ok(expiry)
+    }
+
+    /// Create a new AES-GCM cipher entity.
+    fn create_cipher() -> Result<Aes256Gcm, Error> {
+        // Randomly generate a 256-bit key to be used for encryption/decryption purposes.
+        let mut key = [0u8; KEY_LEN];
+        read_into_buf_from_entropy_pool(&mut key).map_err(Error::EntropyPool)?;
+
+        // Create cipher entity to handle encryption/decryption.
+        Ok(Aes256Gcm::new(Key::from_slice(&key)))
+    }
+
+    /// Make sure to reinitialize the cipher under a new key before reaching
+    /// a count of 2^32 encrypted tokens under the same cipher entity.
+    fn check_encryption_count(&mut self) -> Result<(), Error> {
+        // Make sure no more than 2^32 - 1 tokens are encrypted under
+        // the same encryption key.
+        if self.num_encrypted_tokens == u32::MAX {
+            // Reinitialize the cipher entity under a new key.
+            self.cipher = TokenAuthority::create_cipher()?;
+            // Reset encrypted tokens count.
+            self.num_encrypted_tokens = 0;
+            warn!(
+                "MMDS's token authority entity has been reseeded. \
+                All previously created tokens are now invalid."
+            );
+        }
+
+        Ok(())
+    }
+
+    /// Validate the token time to live against bounds.
+    fn check_ttl(ttl_seconds: u32) -> bool {
+        MIN_TOKEN_TTL_SECONDS <= ttl_seconds && ttl_seconds <= MAX_TOKEN_TTL_SECONDS
+    }
+
+    /// Compute expiry time in seconds by adding the time to live provided
+    /// to the current time measured in seconds.
+    fn compute_expiry(ttl_seconds: u32) -> u64 {
+        // Get time elapsed since UNIX_EPOCH.
+        let now_as_duration = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("System time is before UNIX_EPOCH.");
+        // Convert elapsed time into seconds.
+        let ttl_duration = Duration::from_secs(u64::from(ttl_seconds));
+
+        // Compute expiry by adding ttl value to current time in seconds.
+        // This addition is safe because ttl is verified beforehand and
+        // can never be more than 6h.
+        let expiry_duration = now_as_duration.add(ttl_duration);
+
+        expiry_duration.as_secs()
+    }
+}
+
+/// Structure for token information.
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+struct Token {
+    version: u16,
+    payload_size: u16,
```

Comment:
This was initially added for validation purposes, to make sure the token was generated using the token authority. However, since the `Token` struct also contains a `version` constant field, we can ditch the additional `payload_size` field. I removed it.